### PR TITLE
[SYCLomatc][Intercept-build] Refine intercept-build to work as normal compilation if the toolchain is available

### PR DESCRIPTION
**Pull request by tomflinda**
Signed-off-by: chenwei.sun <chenwei.sun@intel.com>

### CODE REVIEW

**Review by zhiweij1 — APPROVED** [review]

File: clang/tools/scan-build-py/lib/libear/ear.c [hunk 1]
```diff
@@ -1759,156 +1744,261 @@ static void bear_report_call(char const *fun, char const *const argv[]) {
...
-
-  // value 1: means current command line is a compiler command, and the fake obj
-  // file has been created, else ret is set to 0.
-  int ret = 0;
-
-  // To indicate whether the captured cmd is an ar static library command.
-  int is_ar_staticlib_cmd = 0;
+  int is_tool_existence = 0;
+  if (file_exists(argv[0])) {
+    is_tool_existence = 1;
```
Comment (zhimingwang36) [inline clang/tools/scan-build-py/lib/libear/ear.c]: is_tool_available

File: clang/tools/scan-build-py/lib/libear/ear.c [hunk 2]
```diff
@@ -1759,156 +1744,261 @@ static void bear_report_call(char const *fun, char const *const argv[]) {
...
+  }
 
-  const char *command_cp = NULL;
-  size_t it_cp = 0;
-  // (CPATH=;command  args), need remove () around the command
-  int has_parenthesis = 0;
+  if (is_tool_existence) {
+    for (size_t it = 0; it < argc; ++it) {
+      fprintf(fd, "%s%c", argv[it], US);
```
Comment (zhimingwang36) [inline clang/tools/scan-build-py/lib/libear/ear.c:1797]: debug purpose printf?

File: clang/tools/scan-build-py/lib/libear/ear.c [hunk 3]
```diff
@@ -1759,156 +1744,261 @@ static void bear_report_call(char const *fun, char const *const argv[]) {
...
+    //      "/bin/sh -c "/path/to"/bin/clang++ -ccbin ..."
+    const char *pos = find_intercept_compiler(argv[it_cp], compiler_idx);
+
+    int is_stub_need = 0;
+    is_stub_need = (pos != NULL);
+    if (is_stub_need) {
+      ret = 0; // intercept-stub should continue to run.
+
+      // intercept-stub is used to handle the compiler command like
+      // "/bin/sh -c clang++ -c `echo ./`hello.c", it changes the compiler
```
Comment (zhimingwang36) [inline clang/tools/scan-build-py/lib/libear/ear.c]: please keep footprint as small as possible for a PR.

File: clang/tools/scan-build-py/lib/libear/ear.c [hunk 4]
```diff
@@ -1924,88 +2014,9 @@ static void bear_report_call(char const *fun, char const *const argv[]) {
...
-    // coming command "/path/to/libear/intercept-stub -c ./hello.c" will be run,
-    // and the source file name "hello.c" will be captured by intercept.py.
-    argv[it_cp] =
-        replace_binary_name(argv[it_cp], pos, compiler_idx, compiler_array);
-  }
-
-  if (ret == 1 && it_cp == 0 && !is_ar_staticlib_cmd) {
-    exit(0);
+  if(is_tool_existence) {
+    return 0;
```
Comment (zhimingwang36) [inline clang/tools/scan-build-py/lib/libear/ear.c]: it's void function,  so why " return 0"?

File: clang/tools/scan-build-py/lib/libear/ear.c [hunk 5]
```diff
@@ -492,35 +493,14 @@ int eaccess(const char *pathname, int mode) {
...
     return 0;
-  } else if (len > 4 && pathname[len - 1] == 'c' && pathname[len - 2] == 'c' &&
-             pathname[len - 3] == 'v' && pathname[len - 4] == 'n' &&
-             pathname[len - 5] == '/') {
-    // To handle case like "/path/to/nvcc foo.cu ..."
-    return 0;
   }
-  return call_eaccess(pathname, mode);
-}
-
```
Comment (zhimingwang36) [inline clang/tools/scan-build-py/lib/libear/ear.c:503]: I think we should keep this.

File: clang/tools/scan-build-py/lib/libear/ear.c (hunk 2 above)
Comment (tomflinda) [reply clang/tools/scan-build-py/lib/libear/ear.c:1797]: No, it is used to collect options for current command.

File: clang/tools/scan-build-py/lib/libear/ear.c (hunk 3 above)
Comment (tomflinda) [reply clang/tools/scan-build-py/lib/libear/ear.c]: Those codes are in the `else branch` in the code below, logic keeps same as old code
```
  if (is_tool_existence) {
    for (size_t it = 0; it < argc; ++it) {
      fprintf(fd, "%s%c", argv[it], US);
    }
  } else {
    ...
}
```

File: clang/tools/scan-build-py/lib/libear/ear.c (hunk 4 above)
Comment (tomflinda) [reply clang/tools/scan-build-py/lib/libear/ear.c]: It is not void function, shown as follows:
```
static int bear_report_call(char const *fun, char const *argv[]) {
#else
static void bear_report_call(char const *fun, char const *const argv[]) {
#endif // SYCLomatic_CUSTOMIZATION

```

File: clang/tools/scan-build-py/lib/libear/ear.c (hunk 5 above)
Comment (tomflinda) [reply clang/tools/scan-build-py/lib/libear/ear.c:503]: Yes, we keep it, just do a little refinement.

File: clang/tools/scan-build-py/lib/libear/ear.c (hunk 1 above)
Comment (tomflinda) [reply clang/tools/scan-build-py/lib/libear/ear.c]: okay

**Review by zhimingwang36 — APPROVED** [review]

**Pull request merged**